Use grpc.WithBlock() to wait for grpc connections to come up.

avaskys opened the pull request:
We were rolling our own way of doing this via WaitForStateChange on
the connection object. But we could leak that connection object if
we didn't go directly from "connecting" to "idle", as the code
expected. It's cleaner just to use the WithBlock() option and let
grpc determine when we're connected.

Commit:
Use grpc.WithBlock() to wait for grpc connections to come up.

We were rolling our own way of doing this via WaitForStateChange on
the connection object. But we could leak that connection object if
we didn't go directly from "connecting" to "idle", as the code
expected. It's cleaner just to use the WithBlock() option and let
grpc determine when we're connected.

keunwoo commented:
Is there a reasonably self-contained way to write a test that constructs a metricd server and client pair, calls tryConnect, and verifies that the client is in a reasonable state when tryConnect returns (e.g. by trying to do a send)?

Commit:
Added TestClientConnect

This test starts a client and then three seconds later starts a
grpc server for the client to connect to. The expected behavior
is that the connect process blocks until the server is up.

Commit:
Added TestClientConnect

This test starts a client and then three seconds later starts a
grpc server for the client to connect to. The expected behavior
is that the connect process blocks until the server is up.

keunwoo commented on client_test.go:
```diff
@@ -0,0 +1 @@
+package metricd
```
Did this get truncated?

Commit:
Added TestClientConnect

This test starts a client and then three seconds later starts a
grpc server for the client to connect to. The expected behavior
is that the connect process blocks until the server is up.

avaskys commented:
Whoops, yeah looks like I only pushed the blank file. Should be fixed now.

keunwoo commented on client_test.go:
```diff
@@ -0,0 +1,65 @@
+		t.Fatal(err)
+	}
+
+	if rc.conn.State() != grpc.Ready {
```
This is technically racy, for the same reason that the code that you are replacing is racy: by the time you perform this check, the connection may have gone out of ready state due to a transient network disruption.

OTOH, since this connection runs on the loopback interface, and the amount of load going over it is trivial, it is unlikely for this to happen in practice. But on a heavily loaded CI machine, this test could potentially be flaky. I suggest adding a comment acknowledging this possibility.

avaskys replied:
I agree that a race is technically possible here, but given the time scale involved in timing out a network connection, I'd argue that it is orders of magnitude less likely to happen compared to the flaky tests we've seen in the past. Do you know of a strategy for testing this sort of network-based thing that is time-bounded and not potentially flaky?

But yeah I'll certainly add a comment.

Commit:
Added comment in test about potential race

keunwoo commented:
LGTM
